src/twisted/test/test_defgen.py

```diff
@@ -36,58 +28,116 @@ def CRAP():
     return d
 
 
-# NOTE: most of the tests in DeferredGeneratorTests are duplicated
-# with slightly different syntax for the InlineCallbacksTests below.
-
-
 class TerminalException(Exception):
     pass
 
 
-class BaseDefgenTests:
+def deprecatedDeferredGenerator(f):
     """
-    This class sets up a bunch of test cases which will test both
-    deferredGenerator and inlineCallbacks based generators. The subclasses
-    DeferredGeneratorTests and InlineCallbacksTests each provide the actual
-    generator implementations tested.
+    Calls L{deferredGenerator} while suppressing the deprecation warning.
+
+    @param f: Function to call
+    @return: Return value of function.
     """
+    return runWithWarningsSuppressed(
+        [
+            SUPPRESS(
+                message="twisted.internet.defer.deferredGenerator was " "deprecated"
+            )
+        ],
+        deferredGenerator,
+        f,
+    )
+
 
+class DeferredGeneratorTests(unittest.TestCase):
     def testBasics(self):
         """
         Test that a normal deferredGenerator works.  Tests yielding a
         deferred which callbacks, as well as a deferred errbacks. Also
         ensures returning a final value works.
         """
 
-        return self._genBasics().addCallback(self.assertEqual, "WOOSH")
+        @deprecatedDeferredGenerator
+        def _genBasics():
+            x = waitForDeferred(getThing())
+            yield x
+            x = x.getResult()
+
+            self.assertEqual(x, "hi")
+
+            ow = waitForDeferred(getOwie())
+            yield ow
+            try:
+                ow.getResult()
+            except ZeroDivisionError as e:
+                self.assertEqual(str(e), "OMG")
+            yield "WOOSH"
+            return
 
-    def testBuggy(self):
+        return _genBasics().addCallback(self.assertEqual, "WOOSH")
+
+    def testProducesException(self):
         """
-        Ensure that a buggy generator properly signals a Failure
-        condition on result deferred.
+        Ensure that a generator that produces an exception signals
+        a Failure condition on result deferred by converting the exception to
+        a L{Failure}.
         """
-        return self.assertFailure(self._genBuggy(), ZeroDivisionError)
+
+        @deprecatedDeferredGenerator
+        def _genProduceException():
+            yield waitForDeferred(getThing())
+            1 // 0
+
+        return self.assertFailure(_genProduceException(), ZeroDivisionError)
 
     def testNothing(self):
         """Test that a generator which never yields results in None."""
 
-        return self._genNothing().addCallback(self.assertEqual, None)
+        @deprecatedDeferredGenerator
+        def _genNothing():
+            if False:
+                yield 1
```

Comment:
It's best to explicitly inform the `coverage` tool that we don't expect this line to execute.

```suggestion
                yield 1  # pragma: no cover
```